ui/src/App/index.jsx

```diff
@@ -40,7 +41,57 @@ export default class App extends Component {
     introspectionQueryResultData,
   });
 
-  cache = new InMemoryCache({ fragmentMatcher: this.fragmentMatcher });
+  cache = new InMemoryCache({
+    fragmentMatcher: this.fragmentMatcher,
+    /* eslint-disable no-underscore-dangle */
+    dataIdFromObject: object => {
+      const taskId =
+        object.taskId ||
+        ('status' in object && object.status.taskId) ||
+        ('node' in object &&
+          'status' in object.node &&
+          object.node.status.taskId);
+      const status = object.status || ('task' in object && object.task.status);
+      const runs = object.runs || (status && status.runs) || [];
+      const runId = runs.length - 1;
+
+      switch (object.__typename) {
+        case 'TaskDefined': {
+          return `${taskId}/${runId}` || defaultDataIdFromObject(object);
+        }
+
+        case 'TaskPending': {
+          return `${taskId}/${runId}` || defaultDataIdFromObject(object);
+        }
+
+        case 'TaskRunning': {
+          return `${taskId}/${runId}` || defaultDataIdFromObject(object);
+        }
+
+        case 'TaskCompleted': {
+          return `${taskId}/${runId}` || defaultDataIdFromObject(object);
+        }
+
+        case 'TaskFailed': {
+          return `${taskId}/${runId}` || defaultDataIdFromObject(object);
+        }
+
+        case 'TaskException': {
+          return `${taskId}/${runId}` || defaultDataIdFromObject(object);
+        }
+
+        case 'TaskStatus': {
+          return `${taskId}/${runId}` || defaultDataIdFromObject(object);
+        }
+
+        default: {
+          // fall back to default handling
+          return defaultDataIdFromObject(object);
+        }
+      }
+    },
+    /* eslint-enable no-underscore-dangle */
```

Comment:
I have questions :)

I think all of the `|| defaultDataIdFromObject(object)` will never get used, because the left side of `||` will never be false.  Also, this function begins by carefully "guessing" at where taskId and runId might be, in such a way that it will not throw an error for *any* object, even if it might get a bogus taskId or runId or more likely just undefined.  *Then* it consults `__typename` which could have indicated what sort of object it had to begin with.  And all of the branches of the switch do the same thing, so they could probably be coalesced.  So perhaps

```js
  switch (object.__typename) {
    case 'TaskDefined':
    case 'TaskPending':
    case 'TaskRunning':
    case 'TaskCompleted':
    case 'TaskFailed':
    case 'TaskException': {
      const taskId = ..; // wherever a taskId is in these messages
      const runId = ..; // similar
      return `${taskId}/${runId}`;
    }
    case 'TaskStatus': {
      const taskId = ..; // wherever a taskId is in this API response
      const runId = ..; // similar
      return `${taskId}/${runId}`;
    }
  }
```

I can't tell for sure from [the apollo-cache-inmemory docs](https://github.com/apollographql/apollo-client/tree/master/packages/apollo-cache-inmemory) but it looks like the dataId is scoped by the typename, so even if a TaskPending and a TaskStatus both resolve to the same `<taskId>/<runId>`, they won't be treated as the same object in the cache.  And this is probably for the best -- for run 0 there will be a TaskDefined, then TaskPending, then TaskRunning, then TaskResolved message.  All of which would be represented as a TaskStatus with `<taskId>/0` depending on when you fetched it.  So perhaps TaskStatus should not be cached at all (since it's not immutable)?  Or perhaps all of the events should be transmuted into TaskStatus objects?

Reply:
>  Or perhaps all of the events should be transmuted into TaskStatus objects?

I think this approach make things simpler. I'll see what I can do in web-server to have this fixed. Thanks for bringing that up.

Reply:
Questions answered :)